### PR TITLE
Add meaningful message when re declaring default libs

### PR DESCRIPTION
**Pull request by amalp12**
[Issue](https://github.com/gradle/gradle/issues/21328) 

Added appropriate error message to inform users of the possible solutions to the issue.

### Context

[discussion](https://github.com/gradle/gradle/issues/21328)
 
### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java [hunk 1]
```diff
@@ -303,7 +303,8 @@ private void importCatalogFromFile(File modelFile) {
             throw throwVersionCatalogProblemException(getProblemsService(), getProblemsService().getInternalReporter().create(builder ->
                 configureVersionCatalogError(builder, getProblemInVersionCatalog() + "import of external catalog file failed.", CATALOG_FILE_DOES_NOT_EXIST)
                     .details("File '" + modelFile + "' doesn't exist")
-                    .solution("Make sure that the catalog file '" + modelFile.getName() + "' exists before importing it")));
+                    .solution("Make sure that the catalog file '" + modelFile.getName() + "' exists before importing it")
+                    .solution("libs.versions.toml in gradle folder is already taken as default 'libs', please avoid re-declaring it")));
```
[the remark this answers is not in the document]
Comment (amalp12) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java:307]: Sure will do

[the remark this answers is not in the document]
Comment (amalp12) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java:307]: I have tried to replicate this issue in integration test, the build is succeeding, but on my local the build is failing with this issue.

**Comment by amalp12** [comment]
[Here](https://github.com/amalp12/gradle-Issue-21328-reproducible) is the repo to reproduce the issue. 

What is happening is that the init script automatically creates both libs variable and the `libs.version.toml `file.
gradle init --type java-application  --dsl kotlin

### Steps to reproduce 

```
gradle init --type java-application  --dsl kotlin

Enter target Java version (min: 7, default: 21): 17

Project name (default: test): 

Select application structure:
  1: Single application project
  2: Application and library project
Enter selection (default: Single application project) [1..2] 1   

Select test framework:
  1: JUnit 4
  2: TestNG
  3: Spock
  4: JUnit Jupiter
Enter selection (default: JUnit Jupiter) [1..4] 1

Generate build using new APIs and behavior (some features may change in the next minor release)? (default: no) [yes, no] no


> Task :init
Learn more about Gradle by exploring our Samples at https://docs.gradle.org/8.11.1/samples/sample_building_java_applications.html

BUILD SUCCESSFUL in 46s
1 actionable task: 1 executed
```

Append the following to `settings.gradle.kts`

```
dependencyResolutionManagement {
    versionCatalogs {
        create("libs") {
            from(files("gradle/libs.versions.toml"))
        }
    }
}
```

### Build the project 
```

 ./gradlew build
Downloading https://services.gradle.org/distributions/gradle-8.11.1-bin.zip
.............10%.............20%.............30%.............40%.............50%.............60%.............70%.............80%.............90%.............100%
Calculating task graph as no cached configuration is available for tasks: build
[Incubating] Problems report is available at: file:///Users/amal/IdeaProjects/test/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Invalid catalog definition:
  - Problem: In version catalog libs, you can only call the 'from' method a single time.
    
    Reason: The method was called more than once.
    
    Possible solution: Remove further usages of the method call.
    
    For more information, please refer to https://docs.gradle.org/8.11.1/userguide/version_catalog_problems.html#too_many_import_invocation in the Gradle documentation.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 37s
Configuration cache entry stored.

```

### Gradle Verion

```

./gradlew --version

------------------------------------------------------------
Gradle 8.11.1
------------------------------------------------------------

Build time:    2024-11-20 16:56:46 UTC
Revision:      481cb05a490e0ef9f8620f7873b83bd8a72e7c39

Kotlin:        2.0.20
Groovy:        3.0.22
Ant:           Apache Ant(TM) version 1.10.14 compiled on August 16 2023
Launcher JVM:  17.0.13 (Eclipse Adoptium 17.0.13+11)
Daemon JVM:    /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home (no JDK specified, using current Java home)
OS:            Mac OS X 14.5 aarch64


```

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (amalp12) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java]: I have updated the location of the error message

[the remark this answers is not in the document]
Comment (amalp12) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java]: This is the [issue](https://gradle-community.slack.com/archives/C03PU37T8TZ/p1732340903396419) i am facing regarding writing ITs currently, working on it.

**Comment by Vampire** [comment]
It's not only about "redeclaring". Having the default file in the file system counts as `from` call and then doing a `from` call with any value counts as second one, not only when "redeclaring" the default. 

A better fix would probably be to not use the default file if there is an explicit `from`. :-)

**Comment by amalp12** [comment]
Will work on this fix

**Comment by amalp12** [comment]
How about this approach,

I create a boolean 

` private Boolean defaultImportedCatalog;
`

In `DefaultVersionCatalogBuilder`

create another method `fromDefault` which calls from, but also sets the boolean mentioned above to true.

Change [this](https://github.com/amalp12/gradle/blob/4e30e638ee720d00509f988066f3ac5fd90bce9d/subprojects/core/src/main/java/org/gradle/configuration/BuildTreePreparingProjectsPreparer.java#L89) call from` builder.from` to `builder.fromDefault`

Finally change [this](https://github.com/amalp12/gradle/blob/d1ee068b1ee7f62ffcbb549352469307781af72e/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java#L285) check 

from 

`if (importedCatalog == null)
`
to 

`if (importedCatalog == null || this.defaultImportedCatalog)`

**Comment by tresat** [comment]
I responded in the community Slack with help for setting up an integration test.  Give your proposed solution a try and I'll take another look once it's ready and testable.

**Comment by amalp12** [comment]
closing this mr here as a different solution has been suggested

**Comment by amalp12** [comment]
Please take a look at [this](https://github.com/gradle/gradle/pull/31539) MR instead @tresat

**Pull request closed without merging**